### PR TITLE
Add --local-ranks-filter to torchrun: allow logs filtering by rank

### PR DESCRIPTION
**Pull request by xmfan**
Addresses issue https://github.com/pytorch/pytorch/issues/117383

The implementation exposes `--local-ranks-filter` which filters by rank which files we pass to `TailLog` (used in torchrun to determine which logs to output to stdout/stderr)

### with --tee
Currently --tee is implemented as --redirect to file, and streams file to console using `tail`. When --tee is specified, file logs will be unaffected and we will only filter the output to console.

### with --redirect
When --redirect is specified without --tee, nothing is logged to console, so we no-op.

### with neither
When neither --tee or --redirect are specified, torchrun uses empty string "" to indicate logging to console. We intercept this empty string, and redirect it to "/dev/null" to not print to console.

The api also allows a per-rank configuration for --tee and --redirect, and is also supported by this filter implementation.

### without --tee
```
> TORCH_LOGS_FORMAT="%(levelname)s: %(message)s" TORCH_LOGS="graph" torchrun --standalone --nproc_per_node=2 --role rank --local_rank_filter=0 t.py
hello from rank 0 python
DEBUG: TRACED GRAPH
 __compiled_fn_0 <eval_with_key>.0 opcode         name    target                   args       kwargs
-------------  ------  -----------------------  ---------  --------
placeholder    l_x_    L_x_                     ()         {}
call_function  mul     <built-in function mul>  (l_x_, 5)  {}
output         output  output                   ((mul,),)  {}
...
```
### with --tee
```
> TORCH_LOGS_FORMAT="%(levelname)s: %(message)s" TORCH_LOGS="graph" torchrun --standalone --nproc_per_node=2 --role rank --tee 3 --local_rank_filter=0 t.py
[rank0]:hello from rank 0 python
[rank0]:DEBUG: TRACED GRAPH
[rank0]: __compiled_fn_0 <eval_with_key>.0 opcode         name    target                   args       kwargs
[rank0]:-------------  ------  -----------------------  ---------  --------
[rank0]:placeholder    l_x_    L_x_                     ()         {}
[rank0]:call_function  mul     <built-in function mul>  (l_x_, 5)  {}
[rank0]:output         output  output                   ((mul,),)  {}
...
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118562

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Comment by wconstab** [comment]
under the usage example can you clarify the 'with --tee' section to show if/what file it creates in addition to console log??

File: torch/distributed/run.py [hunk 1]
```diff
@@ -548,6 +548,16 @@ def get_args_parser() -> ArgumentParser:
...
     )
 
+    parser.add_argument(
+        "--filter-ranks",
+        "--filter_ranks",
+        action=env,
+        type=str,
+        default="",
+        help="Only show logs from specified ranks in console (e.g. [--filter-ranks 0 1 2] will "
+        "only show logs from rank 0, 1 and 2). When used with --tee, logs will still be saved to files",
```
Comment (wconstab) [inline torch/distributed/run.py:558]: > When used with --tee, logs will still be saved to files
clarify whether _all_ ranks or just filtered ranks files will be saved to files?

**Comment by wconstab** [comment]
suggest you put "--filter-ranks in torchrun" in the topline PR desc

**Review by wconstab — APPROVED** [review]
lgtm! (though i dont know the ins and outs of the torchrun/logging code)

File: torch/distributed/run.py (hunk 1 above)
Comment (xmfan) [reply torch/distributed/run.py:558]: I can add another statement to clarify that this only filters outputs to console, not to files. One confusion is that despite being named `tee_stdouts` and `tee_stderrs`, once they are used in torch/distributed/elastic/multiprocessing/__init__.py:252, they are only used by `TailLogs` to output from file to console

File: torch/distributed/run.py [hunk 2]
```diff
@@ -548,6 +548,16 @@ def get_args_parser() -> ArgumentParser:
         help="Tee std streams into a log file and also to console (see --redirects for format).",
     )
 
+    parser.add_argument(
+        "--filter-ranks",
```
Comment (kurman) [inline torch/distributed/run.py:552]: Overall looks good. Technically this is filtering using local ranks?

Comment (xmfan) [reply torch/distributed/run.py:552]: Yeah, I can change the wording. Although, I don't think we can see non-local ranks logs of a different node?

**Review by wanchaol — APPROVED** [review]
Nice!

Shall we use filter_ranks and delete the TORCH_LOG_RANKS way, or we still keep the later in case user does not use torchrun?

**Comment by xmfan** [comment]
I'm leaning toward deleting the TORCH_LOGS_RANKS, it doesn't filter c++ and logs outside of pytorch. Everyone I've spoken also happen to use elastic launch, and I haven't heard of drawbacks with elastic that would push users away. IMO we should just recommend a single first-party launcher, and continue building on top of it

File: torch/distributed/run.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/run.py:552]: thats a good point, bc anyone trying to use this command from a higher level would need to map from global ranks to local ranks when configuring this arg.  I'm not sure if we intend anyone besides local devgpu users to use this feature.  but if we do, we might want to address global rank usability down the road.

**Comment by xmfan** [comment]
For some reason, MAST precheck flags are being passed through torchrun, and flags starting with "b", "e", "g", "f" will currently fail with `launch.par: error: ambiguous option:`

**Pull request closed without merging**